src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs

```diff
@@ -0,0 +1,389 @@
+        {
+            if (contentHeaderNames.Contains(headerName))
+            {
+                // If this is a "Content" header then we retrieve the value either
+                // from the message content (if available) or the cache (otherwise)
+                if (this.requestMessage.Content != null)
+                {
+                    return string.Join(",", this.requestMessage.Content.Headers.GetValues(headerName));
+                }
+
+                string headerValue;
+                return this.contentHeaderValueCache.TryGetValue(headerName, out headerValue) ? headerValue : string.Empty;
+            }
+            return this.requestMessage.Headers.Contains(headerName) ? string.Join(",", this.requestMessage.Headers.GetValues(headerName)) : string.Empty;
+        }
+
+        /// <summary>
+        /// Sets the value of the header with the given name.
+        /// </summary>
+        /// <param name="headerName">Name of the header.</param>
+        /// <param name="headerValue">Value of the header.</param>
+        public override void SetHeader(string headerName, string headerValue)
+        {
+            if (contentHeaderNames.Contains(headerName))
+            {
+                // If this is a "Content" header then we cache the value (due
+                // to the message content not being set yet) and set it
+                // upon sending the message (when the content will be available)
+                this.contentHeaderValueCache[headerName] = headerValue;
+                return;
+            }
+
+            this.requestMessage.Headers.Remove(headerName);
+            this.requestMessage.Headers.Add(headerName, headerValue);
+        }
+
+        /// <summary>
+        /// Gets the stream to be used to write the request payload.
+        /// </summary>
+        /// <returns>Stream to which the request payload needs to be written.</returns>
+        public override Stream GetStream()
+        {
+            return this.messageStream;
+        }
+
+        /// <summary>
+        /// Abort the current request.
+        /// </summary>
+        public override void Abort()
+        {
+            this.client.CancelPendingRequests();
+        }
+
+        /// <summary>
+        /// Begins an asynchronous request for a System.IO.Stream object to use to write data.
+        /// </summary>
+        /// <param name="callback">The System.AsyncCallback delegate.</param>
+        /// <param name="state">The state object for this request.</param>
+        /// <returns>An System.IAsyncResult that references the asynchronous request.</returns>
+        public override IAsyncResult BeginGetRequestStream(AsyncCallback callback, object state)
+        {
+            var taskCompletionSource = new TaskCompletionSource<Stream>();
+            taskCompletionSource.TrySetResult(this.messageStream);
+            return taskCompletionSource.Task.ToApm(callback, state);
+        }

```

Comment:
Can we consider migrating from `Begin/End` pattern to async/await?

Reply:
@odero  Begin/End are just method names that have been overridden on this particular class.